velox/dwio/parquet/tests/reader/NestedStructureDecoderTest.cpp

```diff
@@ -0,0 +1,218 @@
+    dwio::common::ensureCapacity<vector_size_t>(
+        lengthsBuffer_, kMaxNumValues, pool_.get());
+  }
+
+  void assertStructure(
+      const uint8_t* definitionLevels,
+      const uint8_t* repetitionLevels,
+      int64_t numValues,
+      uint8_t maxDefinition,
+      uint8_t maxRepeat,
+      std::vector<vector_size_t> expectedOffsets,
+      std::vector<vector_size_t> expectedLengths,
+      std::vector<bool> expectedNulls) {
+    int64_t numCollections = NestedStructureDecoder::readOffsetsAndNulls(
+        definitionLevels,
+        repetitionLevels,
+        numValues,
+        maxDefinition,
+        maxRepeat,
+        offsetsBuffer_,
+        lengthsBuffer_,
+        nullsBuffer_,
+        *pool_);
+
+    assertBufferContent<vector_size_t>(
+        offsetsBuffer_, numCollections + 1, expectedOffsets);
+    assertBufferContent<vector_size_t>(
+        lengthsBuffer_, numCollections, expectedLengths);
+    assertNulls(nullsBuffer_, numCollections, expectedNulls);
+  }
+
+ private:
+  template <typename T>
+  void assertBufferContent(
+      BufferPtr actual,
+      size_t actualSize,
+      std::vector<T> expected) {
+    ASSERT_EQ(actualSize, expected.size());
+
+    for (int i = 0; i < actualSize; ++i) {
+      EXPECT_EQ(actual->as<T>()[i], expected[i]);
+    }
+  }
+
+  template <typename T>
+  void printArray(const T* values, int numValues) {
+    std::cout << numValues << std::endl;
+    for (int i = 0; i < numValues; i++) {
+      std::cout << values[i] << " ";
+    }
+    std::cout << std::endl;
+  }
+
+  void
+  assertNulls(BufferPtr actual, size_t actualSize, std::vector<bool> expected) {
+    ASSERT_EQ(actualSize, expected.size());
+
+    auto nullsBuffer = actual->as<uint64_t>();
+    for (int i = 0; i < actualSize; ++i) {
+      EXPECT_EQ(bits::isBitSet<uint64_t>(nullsBuffer, i), expected[i]);
+    }
+  }
+
+  std::unique_ptr<memory::MemoryPool> pool_;
+  BufferPtr offsetsBuffer_;
+  BufferPtr lengthsBuffer_;
+  BufferPtr nullsBuffer_;
+};
+
+// ------------------------
+// [1, 2, 3, 4]
+// [2,3]
+// [null, 3, null]
+// NULL
+// [null]
+// [null, null]
+// [7, 8, 9]
+// [8]
+// [null]
+TEST_F(NestedStructureDecoderTest, oneLevel) {
+  uint8_t defs[] = {3, 3, 3, 3, 3, 3, 2, 3, 2, 0, 2, 2, 2, 3, 3, 3, 3, 2};
+  uint8_t reps[] = {0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0};
+  std::vector<vector_size_t> expectedOffsets(
+      {0, 4, 6, 9, 9, 10, 12, 15, 16, 17});
+  std::vector<vector_size_t> expectedLengths({4, 2, 3, 0, 1, 2, 3, 1, 1});
+  std::vector<bool> expectedNulls(
+      {false, false, false, true, false, false, false, false, false});
+
+  assertStructure(
+      defs, reps, 18, 1, 1, expectedOffsets, expectedLengths, expectedNulls);
+}
+
+//---------------------------
+//[[1], [2, 3]]
+//[[2, null], [2]]
+//[[3, null], null, [4, 5]]
+//[[4, null], null, null]
+// NULL
+//[null, [null], null, [6]]
+//[[7], null, [7]]
+//[null, [8, 9]]
+TEST_F(NestedStructureDecoderTest, twoLevels) {
```

Comment:
I assume `readOffsetAndNulls` handles intermediate levels also and not just the leaves?
Can we add a test here for
`assertStructure(
      defs, reps, 26, 1, 1, expectedOffsets, expectedLengths, expectedNulls); ?

Reply:
@majetideepak I added firstLevelInTwo to the existing secondLevelInTwo test, which was to calculate the first level array offsets and nulls for ARRAY<ARRAY<INT>>. It has maxDef 3 and maxRepeat 2. It's already the top level array. The second level array has maxDef 5 and maxRepeat 2. maxDef 1 and maxRepeat 1 is for the top level rows and it's not the concern here. But I will for sure add more tests in near future.